src/libPMacc/include/particles/IdProvider.hpp

```diff
@@ -0,0 +1,161 @@
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "Environment.hpp"
+#include "eventSystem/EventSystem.hpp"
+#include "algorithms/reverseBits.hpp"
+#include "nvidia/atomic.hpp"
+#include "memory/buffers/HostDeviceBuffer.hpp"
+
+namespace PMacc {
+
+    namespace IdDetail {
+
+        __device__ uint64_cu nextId;
+
+        __global__ void setNextId(uint64_cu id)
+        {
+            nextId = id;
+        }
+
+        template<class T_Box>
+        __global__ void getNextId(T_Box boxOut)
+        {
+            boxOut(0) = nextId;
+        }
+
+        template<class T_Box, class T_GetNewId>
+        __global__ void getNewId(T_Box boxOut, T_GetNewId getNewId)
+        {
+            boxOut(0) = getNewId();
+        }
+
+    }  // namespace IdDetail
+
+    /**
+     * Provider for globally unique (even across ranks) ids
+     * Implemented for use in static contexts
+     */
+    class IdProvider
+    {
+    public:
+        /** Initializes the state so it is read for use
+         */
+        static void init()
+        {
+            const uint64_t globalUniqueStartId = getStartId();
+            setNextId(globalUniqueStartId);
+            // Instantiate kernel
+            getNewIdHost();
+            // Reset to start value
+            setNextId(globalUniqueStartId);
+        }
+
+        /** Sets the next id to a given value (e.g. after a restart)
+         */
+        static void setNextId(const uint64_t nextId)
+        {
+            __cudaKernel(IdDetail::setNextId)(1, 1)(nextId);
+        }
+
+        /** Returns the next id without changing it (e.g. for saving)
+         */
+        static uint64_t getNextId()
+        {
+            HostDeviceBuffer<uint64_cu, 1> nextIdBuf(DataSpace<1>(1));
+            __cudaKernel(IdDetail::getNextId)(1, 1)(nextIdBuf.getDeviceBuffer().getDataBox());
+            nextIdBuf.deviceToHost();
+            return nextIdBuf.getHostBuffer().getDataBox()(0);
+        }
+
+        /** Functor that returns a new id */
+        struct GetNewId
+        {
+            DINLINE uint64_cu operator()() const
+            {
+                return getNewId();
+            }
+        };
+
+        /** Function that returns a new id */
```

Comment:
pls add a bit more documentation, here the value is changed so in contrast to `getNextId()` this changes the state.

(Yes, in a non static design the difference would be obvious with `const` and non-const member functions.)